ipaserver/plugins/servicedelegation.py

```diff
@@ -212,22 +223,37 @@ def post_callback(self, ldap, completed, failed, dn, entry_attrs,
         failed[self.principal_failedattr] = {}
         failed[self.principal_failedattr][self.principal_attr] = []
         names = options.get(self.member_names[self.principal_attr], [])
-        ldap_obj = self.api.Object['service']
+        basedn = self.api.env.container_accounts + self.api.env.basedn
         if names:
             for name in names:
                 if not name:
                     continue
-                name = normalize_principal(name)
-                obj_dn = ldap_obj.get_dn(name)
+
+                try:
+                    n = kerberos.Principal(name, realm=self.api.env.realm)
+                except ValueError:
+                    raise errors.ValidationError(
+                        name='principal', reason=_("Malformed principal"))
+
+                if len(n.components) == 1 and (n.components[0][-1] != '$'):
+                    nprincipal = 'host/' + unicode(n)
+                else:
+                    nprincipal = unicode(n)
```

Comment:
I'm not sure the unicode bit is necessary, this is going to be python3-only right? Or are you being considerate of backports?

Reply:
Yes, I want to not preclude this possibility. Additionally, the rest of the module is py2-friendly and I am keeping the status quo here. When we'd decide to clean everything of the py2 compatibility, it will need to be done for the whole file anyway.

Reply:
+1